[#4] 카테고리 생성 기능 구현

gomster96 opened the pull request:
### Description
카테고리를 생성할 수 있는 기능을 서비스까지 만들었다. 
(Controller는 관리자 페이지와 같이 다음 이슈에서 추가 예정) 

- category 추가 기능 구현 (단일 이름 추가, parent 카테고리가 있을 때 추가) 
- parent 연관관계 편의 메소드 작성 
- category 단일 추가 테스트, 부모와 함께 추가 테스트 

### Trouble Shooting


### ETC
- 깃 크라켄을 처음 사용해보면서 이상한 커밋메세지나, 잘못된 커밋들이 올라가서 이부분 죄송합니다.

gomster96 committed:
# <type>: <subject> 꼴로 작성한다 

feat : 카테고리 다대다를 위해 ShopCategory 생성

- Shop에서도 Category를 다수 가지기 위해 ShopCategory를 만들어, Category에서 Shop을, Shop에서 Category를 가질수 있도록 하였다.

gomster96 committed:
# <type>: <subject> 꼴로 작성한다 

feat : Category Exception 생성 
##### Subject 50 characters ################# -> |


# Body Message
######## Body 72 characters ####################################### -> |

# Issue Tracker Number or URL
# 특정 이슈를 참조하는 커밋일 경우에 해당 이슈를 참조한다. 


# --- COMMIT END ---
# Type can be
#   feat    : 새로운 기능에 대한 커밋
#   fix     : 버그 수정에 대한 커밋
#   refactor: 코드 리팩토링에 대한 커밋
#   style   : 코드 스타일 혹은 포맷 등에 관한 커밋
#   build   : 빌드 관련 파일 수정에 대한 커밋
#   docs    : 문서 수정에 대한 커밋
#   test    : 테스트 코드 수정에 대한 커밋
#   chore   : 그 외 자잘한 수정에 대한 커밋
# ------------------
# Remember me ~
#   Capitalize the subject line
#     제목줄은 대문자로 시작한다.
#   Use the imperative mood in the subject line
#     제목줄은 명령어로 작성한다.
#   Do not end the subject line with a period
#     제목줄은 마침표로 끝내지 않는다.
#   Separate subject from body with a blank line
#     본문과 제목에는 빈줄을 넣어서 구분한다.
#   Use the body to explain what and why vs. how
#     본문에는 "어떻게" 보다는 "왜"와 "무엇을" 설명한다.
#   Can use multiple lines with "-" for bullet points in body
#     본문에 목록을 나타낼때는 "-"로 시작한다.
# ------------------

gomster96 committed:
feat : Category Service 구현

- Category service 구현 
- isCategoryExist 로 findByName Exception 처리해주었다. 
- add 할때 name 과 parentName 있을 때 를 구분해서 구현하였다.

gomster96 committed:
feat: CategoryRepository 구현

- CategoryRepository를 구현할 떄 JpaRepository를 extend하여 사용하였다. 
- findByName 메서드를 생성하여, 이름에 따라 해당 Category 가 존재하는지 파악하기 위해 구현하였다.

gomster96 committed:
refactor: categoryService 리팩토링

- builder에서 굳이 중복되는 코드를 만들었기에 지웠음

gomster96 committed:
feat: Category 생성자에 Builder 추가 및 name Validation

- 생성자에 Builder를 추가하면 굳이 생성자를 다 쓰지 않아도 다른 생성자들도 Builder로 사용할 수 있다.
- 또한 생성자에 Builder를 추가했기에 name에 validation을 쓸 수 있게 되었다.

gomster96 committed:
test: CategoryNameTest 구현

- Category domain 을 생성할 때 빈 문자열이 이름으로 들어오면 exception이 발생하도록 테스트를 구현하였다.

gomster96 committed:
refactor: CategoryRequstDto 생성

- 기존의 코드는 parent, name 으로 String이 두개가 입력되는데, 두개면 dto가 생성되는게 나을 것 같아서 dto를 만들어 주었다.

gomster96 committed:
fix: categoryRequestDto를 사용하여 create 할때 오류 수정

- 실수로 new Category() 라고 적은 것을 newCategory로 바꾸어 오류를 수정하였다.

gomster96 committed:
test : CategoryName 이 없을 경우 테스트

- Category를 등록할 때 이름이 빈 문자열이 들어올 경우 Exception이 발생하도록 테스트를 구현하였다.

gomster96 committed:
test[#4] : 카테고리 Integration 테스트 

- 최상단 카테고리 생성 테스트
- 부모와 함께 생성되는 카테고리 생성 테스트
- 카테고리ㅡㄹ 부모와 함께 생성할 때 부모가 없는 부모면 예외 발생

gomster96 committed:
refactor[#4] : 주석제거 및 indentation 맞춤

gomster96 committed:
refactor[#4] : 커밋에 추가하지 않았던 파일들 추가

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,47 @@
 
     @ManyToOne
     private Category parent;
 
```
fetch type을 lazy로 추가해주면 좋을 거 같아요!!

KJunseo replied:
같은 의견입니다

gomster96 replied:
아 감사합니다. 제가 잊고있었네요 변경해서 다시 올리겠습니다~~

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/shop/domain/category/ShopCategory.java:
```diff
@@ -0,0 +1,19 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    private Shop shop;
+
+    @ManyToOne
+    private Category category;
```
fetch type을 lazy로 추가해주면 좋을 거 같아요!!

gomster96 replied:
아 감사합니다. 제가 잊고있었네요 변경해서 다시 올리겠습니다~~

KJunseo replied:
아직 Lazy가 반영되지 않은 것 같아요 ㅠㅠ

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/shop/domain/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    @Query("select c from Category c where c.name = :user")
+    Optional<Category> findByName(@Param("user") String name);
```
여기서 user parameter는 실제로 user name이 들어오는 건가요??
만약 아니라면, 의미 있는 parameter 명으로 바꾸는 게 좋을 거 같아요!!

gomster96 replied:
음...! 흠 user name이 들어오는게 절대로 아닙니다..! 카테고리 name인데 잘못했군요..! 감사합니다 변경하겠습니다.

PyeongGangKim commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+        String categoryRequestName = "테스트숙소카테고리";
+
+        //when
+        Long id = categoryService.create(categoryRequestName);
+        //then
+        Assertions.assertThat(id).isNotNull();
```
id를 사용해서 category를 찾아와서, category name을 비교하는 거는 어떨까요?

gomster96 replied:
저는 이게 제대로 만들어 졌다면? id 가 생성되기 때문에 이것이 notNull인지를 파악함을 통해 제대로 생성되었음을 확인했는데, 말씀하신대로 id로 찾아서 가져오면 뭔가 한번 더 확인하는 느낌이 아닐까요?

gomster96 replied:
아 근데 지금 생각해보니 제대로 된 name으로 잘 만들어졌는지 까지 확인하는 부분이 필요하겠네요...! 감사합니다 반영하겠습니다 ㅎㅎ

PyeongGangKim commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+                                                                  .parent("테스트식당")
+                                                                  .build();
+        //        //when
+        Long id = categoryService.create(categoryRequestDto);
+        System.out.println(id);
+        //        //then
+        Assertions.assertThat(id).isNotNull();
```
id를 사용해서 category를 찾아와서, category name을 비교하는 거는 어떨까요?

KJunseo replied:
생성한 카테고리를 repository로 찾아와서 실제로 부모연관관계가 잘 들어갔는지 확인해보면 좋을 것 같네요

PyeongGangKim reviewed (changes requested):
코드 보면서 궁금한 점을 달아놨습니다!!

KJunseo commented on backend/src/main/java/com/handong/rebon/exception/ExceptionResponse.java:
```diff
@@ -0,0 +1,2 @@
+package com.handong.rebon.exception;public class ExceptionResponse {
+}
```
pr전에 포맷팅 한 번 적용하면 좋을 것 같아요.
다른 exception들도 한 번씩 확인해 보면 좋을 것 같네요

```suggestion
package com.handong.rebon.exception;

public class ExceptionResponse {
}
```

gomster96 replied:
제가 코드들을 다 push해서 올리지 않았었네요.....!. 이런 실수해서 죄송합니다! 깃 크라켄 쓰다가 적용이 잘못된줄 알고 안올렸나봐요! 수정하겠습니다

KJunseo commented on backend/src/main/java/com/handong/rebon/exception/category/CategoryExistException.java:
```diff
@@ -0,0 +1,2 @@
+package com.handong.rebon.exception.category;public class CategoryExistException {
+}
```
CategoryException을 extends 해야하지 않을까요?

gomster96 replied:
제가 코드들을 다 push해서 올리지 않았었네요.....!. 이런 실수해서 죄송합니다! 깃 크라켄 쓰다가 적용이 잘못된줄 알고 안올렸나봐요! 수정하겠습니다

KJunseo commented on backend/src/main/java/com/handong/rebon/exception/category/CategoryNameException.java:
```diff
@@ -0,0 +1,9 @@
+
+import org.springframework.http.HttpStatus;
+
+public class CategoryNameException extends CategoryException{
+    public CategoryNameException(){
```
```suggestion
public class CategoryNameException extends CategoryException {
    public CategoryNameException() {
```

gomster96 replied:
제가 코드들을 다 push해서 올리지 않았었네요.....!. 이런 실수해서 죄송합니다! 깃 크라켄 쓰다가 적용이 잘못된줄 알고 안올렸나봐요! 수정하겠습니다

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,53 @@
+        return newCategory.getId();
+    }
+
+    private void isCategoryExist(String name) {
+        Optional<Category> category = categoryRepository.findByName(name);
+        if (category.isPresent())
+            throw new CategoryExistException();
+    }
```
db에 값이 있는지를 확인해야하기 때문에 이 과정은 필요할 것 같네요.
`if (categoryRepository.existsByName(name))` 이런식으로 처리해도 될 것 같습니다.

gomster96 replied:
와우....! 이런 더 깔끔한 방법이 있네요 바로 반영하였습니다 감사합니다 ㅎㅎ 많이 배웁니다 👍

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    @Query("select c from Category c where c.name = :user")
+    Optional<Category> findByName(@Param("user") String name);
```
@Query문을 굳이 붙여준 이유가 있으실까요?
`Optional<Category> findByName(String name);` 만 적어줘도 똑같이 동작하지 않나요?

gomster96 replied:
아 이렇게 할수가 있네요! 흠 JpaRepository에 대해서 더 공부해봐야겠습니다 감사합니다~~ 👍

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,53 @@
+        return newCategory.getId();
+    }
+
+    public Long create(CategoryRequestDto categoryRequestDto) {
```
이 부분이 상위 카테고리에 속한 자식 카테고리들을 생성하는 메서드로 보이는데요.
부모 카테고리의 경우 name이 아닌 부모 카테고리의 id를 사용하는 게 좋지 않을까요?
개인적으로는 name 중복처리를 해준다고 해도 id같이 원래 고유한 값을 사용하는 게 더 좋아보여요

KJunseo replied:
지금은 카테고리의 이름이 중복되었는지를 체크하고 있는것 같은데요.
최상위 카테고리만 다르면 이름이 같은 경우도 있을 수 있지 않을까싶은 생각이 살짝 들긴하네요.

ex) 식당 -  인기, 한식, 중식
       카페 - 인기, 프랜차이즈, 개인카페

gomster96 replied:
부모가 있는경우 카테고리의 이름은 중복이 안되게 하는게 맞는것 같습니다..! 

흠 id 값을 아예 requestDto에서 받아오도록 한다고 생각하고 구현하면 괜찮을까요?

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,53 @@
+        Category parent = categoryRepository.findByName(categoryRequestDto.getParent())
+                                            .orElseThrow(CategoryNoParentException::new);
+
+        Category newCategory = Category.builder()
+                                       .name(categoryName)
+                                       .parent(parent)
+                                       .build();
```
자식 카테고리에 부모를 할당해주고, 해당 부모의 children에도 자식을 추가해주는 게 좋을 것 같네요. jpa convenient method 에 대해 찾아보면 좋을 것 같습니다.

추가적으로 Categories에 cascade 옵션을 줘도 될거 같아요

gomster96 replied:
연관관계편의메소드 바로 적용하였습니다 👍  자꾸 성장하게해주셔서 감사합니다

KJunseo commented on backend/src/test/java/com/handong/rebon/unit/category/CategoryNameTest.java:
```diff
@@ -0,0 +1,21 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class CategoryNameTest {
```
테스트 클래스엔 굳이 public을 주지 않아도 되는 걸로 알고 있어요. 실제로 단축키로 테스트를 생성하면 class만 붙어서 생성됩니다!

이 클래스이름을 CategoryNameTest라고 지은 이유가 있으실까요? 후에 카테고리 내에 로직이 추가될 경우 CategoryTest가 더 적절해보입니다.

KJunseo commented on backend/src/test/java/com/handong/rebon/unit/category/CategoryNameTest.java:
```diff
@@ -0,0 +1,21 @@
+    @ValueSource(strings = {"", " "})
+    @DisplayName("카테고리의 이름이 비어있거나 공백으로 들어온 경우 등록할 수 없다.")
+    void blankCategoryName(String name) {
+        Assertions.assertThatThrownBy(() -> Category.builder().name(name).build())
+                  .isInstanceOf(CategoryNameException.class);
```
given-when-then 패턴을 찾아보시면 좋을 것 같아요.
가독성이 훨씬 좋아질 것 같네요.

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+                                                                  .name("테스트한식")
+                                                                  .parent("테스트식당")
+                                                                  .build();
+        //        //when
```
주석이 이상하게 달렸네요 😄

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+                                                                  .build();
+        //        //when
+        Long id = categoryService.create(categoryRequestDto);
+        System.out.println(id);
```
print 작업을 남겨두신 이유가 있나요??

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+                                                                  .name("테스트한식")
+                                                                  .parent("존재하지않는카테고리")
+                                                                  .build();
+        //        //when
+        //        //then
```
😄

KJunseo reviewed (changes requested):
제 의견을 적어봤습니다! 코멘트 부탁드릴게요. 확인해보고 다시 리뷰 요청 주세요

+ clone받아서 돌려보려 했는데 컴파일 에러가 나네요 ㅠㅠ(exception 부분 때문인 것 같습니다)

gomster96 committed:
refactor[#4] : 깃 크라켄 관련 문제 해결

- 깃 크라켄 사용하며 파일들이 자동 커밋으로 올라가는 문제들이 발생했어서 해당 문제를 해결하기 위해 일단 모든 파일을 커밋하고 푸시합니다.

gomster96 committed:
reafactor[#4] : domain에 lazy loading, access protected 추가 

- OneToMany 관계의 경우 FetchType.Lazy 옵션을 넣어주었다. 
- NoargsConstructor의 경우 access level 을 protected로 해주었다.

gomster96 committed:
refactor[#4] : findByName 바꾸기

- JpaRepository에서 제공하는 방법으로 굳이 @Query를 붙일 필요없이 할 수 있음을 알게되어 코드를 교체하였다.

gomster96 committed:
feat[#4] : 카테고리 생성 연관관계 편의 메서드 구현 

- CategoryService 에 연관관계 편의 메서드를 생성하여 parent를 생성함과 동시에 child까지 처리해주는 방법을 적용하였다.

gomster96 committed:
feat[#4] : cascade 옵션 적용

- Categories의 경우 부모가 없어지면 자식도 다 삭제되어야하므로, 또는 저장될 때 자식도 같이 저장되어야하므로 cascade옵션을 주었다.

gomster96 committed:
feat[#4] : 카테고리 생성 및 integration test 완료 

- given when then 방식으로 테스트하기위해 코드 구조를 변경 
- 기존에 category를 생성할 때 children을 넣어주지않았기 때문에 이제 넣어줌

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,58 @@
+                                       .children(new Categories())
+                                       .build();
+        categoryRepository.save(newCategory);
+        return newCategory.getId();
```
이렇게 해줘도 되겠지만 명시적으로 반환값을 사용하는건 어떻게 생각하시나요?
```suggestion
        Category savedCategory = categoryRepository.save(newCategory);
        return savedCategory.getId();
```

gomster96 replied:
아! 명시적으로 하는게 좋은것 같습니다 반영한줄알았는데 안되어있군요 수정하겠습니다 ㅎㅎ

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,58 @@
+    private void isCategoryExist(String name) {
+        Optional<Category> category = categoryRepository.findByName(name);
+        if (category.isPresent())
+            throw new CategoryExistException();
```
if문 내부가 한 줄이어도 일관성있게 {}를 넣어주는게 가독성 면에서 좋아보입니다.

KJunseo replied:
추가적으로 값 존재 여부 검증을하는데 exist가 아닌 find 를 하신 이유가 궁금합니다!

gomster96 replied:
아래 커밋에서 exist로 반영하였습니다! 저도 앞으로는 {}를 쓰도록 하겠습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,58 @@
+        return newCategory.getId();
+    }
+
+    private void isCategoryExist(String name) {
```
is~ 로 시작되는 메서드명은 반환값이 boolean일 것 같은 느낌을 주는 것 같아요

gomster96 replied:
음 맞네요 checkCategoryExist 이렇게 메소드이름 바꾸는게 나을것 같군요 감사합니다 .!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,58 @@
+
+    private final CategoryRepository categoryRepository;
+
+    public Long create(String name) {
```
읽기 쿼리와 쓰기 쿼리가 함께 발생하여 `@Transactional`을 붙여줘야 될 것 같습니다.

gomster96 replied:
음 그렇군요 `@Transactional` 어노테이션 관련해서 공부하고 추가해 놓도록 하겠습니다. 감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,29 @@
 public class Categories {
 
-    @OneToMany(mappedBy = "parent")
+    @OneToMany(mappedBy = "parent", fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
```
`@OneToMany`는 기본값이 Lazy라서 굳이 적어주지 않아도 될 것 같아요! 만약 명시적으로 표시하고 싶었던 의도라면 모든 엔티티에 똑같이 해줘야 일관성있을 것 같네요

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,29 @@
 
-    @OneToMany(mappedBy = "parent")
+    @OneToMany(mappedBy = "parent", fetch = FetchType.LAZY, cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
     private List<Category> categories;
```
collection 타입의 경우 null문제 때문에 바로 초기화 해주는 것이 좋을 것 같네요!
이 부분 말고도 다른 엔티티 부분도 같이 확인해주시면 좋을 것 같습니다.

참고하시면 좋을 것 같습니다! https://velog.io/@yebali/Spring-JPA-Entity%EC%84%A4%EA%B3%84-%EC%8B%9C-%EC%A3%BC%EC%9D%98%EC%A0%90

gomster96 replied:
오... 정말 감사합니다! 생성자를 굳이 주지 않고 저렇게 하는게 훨씬 좋았겠네요..! 엔티티에 바로 new를 할 수 있는걸 몰랐네요 ㅎㅎ 감사합니다 👍

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,50 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@Getter
+@Setter
```
엔티티에 setter가 필요한가요?

gomster96 replied:
생각해보니 범용적인 @setter보다는 그냥 setter 메소드를 하나 만드는게 나을것 같습니다 감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,50 @@
+    }
+
+    public void addChildCategory(Category category){
+        this.getChildren().addChild(category);
```
굳이 getChildren하지 않고 children에 접근할 수 있지 않나요??
```suggestion
        this.children.addChild(category);
```

gomster96 replied:
음 사실 맞습니다. 제가 getter를 사용하는것이 직접 필드에 접근하는 것보다 안전하다는 말을 들어서 근데 어차피 클래스내부라 이때의 경우가 아니였던거 같습니다. 수정하겠습니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,50 @@
+
+    public void addChildCategory(Category category){
+        this.getChildren().addChild(category);
+        category.setParent(this);
```
하는 일이 setter와 동일하더라도 setter는 범용적인 이름이라서 setter라는 네이밍 보다 별도의 네이밍으로 메서드를 만드는 것이 좋아보여요.

gomster96 replied:
https://www.inflearn.com/questions/16235 그러네요 찾아보니 좀더 의미있는 변경 메서드 이름을 사용하는것이 좋겠군요! 

connectParent 라고 이름 짓겠습니다!

gomster96 committed:
refactor[#4] : existsByName으로 refactoring

- 존재여부를 파악할때 JpaRepository에서 제공해주는 existsByName을 만들어서 더 가독성 좋은 코드로 만들었다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void isCategoryExist(String name) {
+        if(categoryRepository.existsByName(name)) throw new CategoryExistException();
```
if문 한 줄이어도 {}를 붙여주는 게 일관성있고 가독성 면에서도 더 좋을 것 같아요
```suggestion
        if(categoryRepository.existsByName(name)) {
            throw new CategoryExistException();
        }
```

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,50 @@
+    private List<ShopCategory> shopCategories;
 
     @ManyToOne
     private Category parent;
```
fetchType Lazy로 주면 좋을 것 같네요

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,77 @@
+                                                                  .name("테스트한식")
+                                                                  .build();
+        //when
+        final RuntimeException exception = assertThrows(RuntimeException.class, () -> categoryService.create(categoryRequestDto));
+        //then
+        assertThat(exception.getMessage()).isEqualTo("저장하려는 카테고리의 부모 카테고리가 존재하지 않습니다.");
```
RuntimeException 말고 실제 뿌려주는 커스텀 Exception으로 검증해도 좋을 것 같네요

KJunseo replied:
여기만 final 붙이신 특별한 이유가 있나요?

gomster96 replied:
사실 큰 이유는 없고 뭔가 한번 발생한 exception 자체를 상수로 넣어주는건 어떤가 싶어서 넣었습니다. 

약간 자바스크립트에서는 일반적으로 변하지 않을 값에 대해서는 앞에다가 const라고 붙여주는데 근데 이거에 경우는 값타입이 아니라 레퍼런스타입일테니 final을 붙이는게 큰 의미를 부여하지 않을 것 같네요..! 수정하는게 나을것 같습니다.

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryIntegrationTest.java:
```diff
@@ -0,0 +1,77 @@
+
+@SpringBootTest
+@Transactional
+public class CategoryIntegrationTest {
```
부모 카테고리 생성시 이미 존재하는 카테고리를 생성하는 경우에 대한 테스트도 있으면 좋겠네요

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,56 @@
+
+        Category newCategory = Category.builder()
+                                       .name(categoryName)
+                                       .children(new Categories())
```
개인적으로는 여기서 new Categories를 넣어주는 게 조금 어색하게 느껴져요. 내부에서 처리할 수도 있을 것 같네요

gomster96 replied:
아 어차피 기능이 같다면, 내부적으로 생성자에 넣어주는 방법이 있겠네요. 감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,56 @@
+        return newCategory.getId();
+    }
+
+    public Long create(CategoryRequestDto categoryRequestDto) {
```
식당 카테고리 자식 카테고리로 한식이 있는 상태에서, 다시 식당 카테고리에 한식을 자식으로 생성하면 어떻게 되나요? addChildCategory를 할 때 별도의 예외처리가 없는 것 같아요. + 이 경우에 대한 테스트 코드도 작성하면 좋겠네요

KJunseo reviewed (changes requested):
고생하셨습니다! 좀 더 리뷰할게 보여서 의견 적어봤습니다! 다시 확인하고 리뷰 요청해주세요

+ 사용하지 않는 import 들이 많이 올라와있는 것 같아요. commit전에 단축키를 사용해서 지워주시는게 좋을 것 같아요!

gomster96 committed:
refactor[#4] : 카테고리 fetchType Lazy변경, categoryService 메소드명 변경

- cetegoryService에서 isExistName은 무엇인가 boolean값을 리턴할거 같은 네이밍이기에 checkExistName으로 바꾸었다.
- 다대일 관계의 경우 default가 즉시로딩이므로 이를 지연로딩으로 바꿔주고, 일대다관계인경우 default가 지연로딩이므로 적었던걸 다시 지워주었다.

gomster96 committed:
refactor[#4] : LazyLoading 추가

gomster96 committed:
refactor[#4] : 엔티티 내부의 collection에 new로 초기화

gomster96 committed:
refactor[#4] : 코드 일관성 유지

- if문에 한줄이라고 괄호를 추가하여 코드의 일관성을 유지시켰다.
- 기존의 builder를 사용할때 category를 넣어줌에 있어서 new Categories()를 넣어주었는데, builder에 new가 들어가는 부분이 부적절하다고 판단하여, builder생성자 내부에서 알아서 new 를 해줄 수 있도록 코드를 변경하였다.

gomster96 committed:
refactor[#4] : 코드 스타일 통일을 위해 굳이 불필요한 getter 사용 제거

gomster96 committed:
test[#4] : 부모 카테고리 생성 시 이미 존재하는 카테고리라면 예외 발생 

- 부모 카테고리의 경우 이름의 유일성을 보장해야하기 때문에 예외에 관련한 테스트를 진행하였다.

gomster96 committed:
feat[#4] : 부모가 존재하는 카테고리를 추가할 때 이름 중복여부 검사 체크

- 부모가 같은 경우 카테고리는 중복여부를 검사해야하기 때문에 해당 기능을 추가하였다.

gomster96 committed:
test[#4] : 같은 부모 카테고리 추가시, 중복일 경우 예외발생 테스트

- 같은 부모인 동일 이름의 데이터를 추가시, exception이 발생하는지의 여부를 테스트하였다.

gomster96 committed:
refactor[#4] : 코드 통일을 위한 모든 파일 포멧팅 적용

- Intellij에서 지원하는 코드 포멧터를 모든 파일에 적용하였다.
- 사용하지 않는 import 문을 제거해주었다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -21,16 +16,16 @@
     private final CategoryRepository categoryRepository;
 
     public Long create(String name) {
```
`@Transactional`이 반영되지 않는 것 같아요!

gomster96 replied:
`@Trascational` 더 공부하고 반영하겠습니다!

gomster96 committed:
refactor[#4] : checkDuplicateCategory 메소드 Categories로 이동

- categories에서만 확인하면 되는 기능이기 때문에 Category 도메인에 있기 보다는 Categories에 위치하는 것이 조금 더 알맞다고 판단하여 위치를 변경시켰습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,52 @@
+
+        String categoryName = categoryRequestDto.getName();
+
+        Category newCategory = Category.builder()
+                                       .name(categoryName)
+                                       .build();
```
이름만으로 생성을 하는데 굳이 빌더패턴을 사용해야할까요? 어떻게 생각하시나요?

gomster96 replied:
아 동의합니다! 잊고 있었네요 , 사실 Category를 생성하는 방법이 name으로 밖에 없다면 그냥 생성자를 만드는게 맞는것 같습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,22 @@
+    @OneToMany(mappedBy = "parent", cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
+    private List<Category> categories = new ArrayList<>();
+
+    public void addChild(Category category) {
+        this.getCategories().add(category);
```
이부분도 getter가 아닌 필드를 바로 이용해도 될 것 같네요

gomster96 replied:
제가 커밋단위를 잘 써야했는데 나중 커밋에 반영되었습니다! 감사합니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,65 @@
 
-    @Embedded
-    private Shops shops;
+    @OneToMany(mappedBy = "category")
+    private List<ShopCategory> shopCategories;
```
collection 초기화 해주시면 좋을거 같아요

gomster96 replied:
이게 참 커밋올리기전에 더 신중했어야하는데, 알고있는데 반영을 안했네요! 반성하겠습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,65 @@
+    }
+
+    private void checkDuplicateCategory(Category category) {
+        this.children.getCategories().forEach(child -> checkSameName(child, category));
```
categories에서 List<Category>를 꺼내는 것 보다 categories에게 메세지를 보내는 건 어떤가요? categories의 책임이라는 생각이 들어요.

gomster96 replied:
혹시 여기부분 저도 똑같이 생각해서 가장 마지막 커밋에 바꾸어서 반영했는데 괜찮겠죠???

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -2,22 +2,65 @@
+    private void checkSameName(Category child, Category category) {
+        if (child.getName().equals(category.getName())) {
+            throw new CategoryExistException();
+        }
```
getter는 가능하면 응답값으로 값을 내보낼 때만 쓰는게 좋다고 생각합니다.
Category에게 메세지를 보내 처리할 수 있지 않을까요?

gomster96 replied:
getter를 응답값으로만 쓸 수 있도록 통일성을 주는것이 좋을것 같네요. 잘 생각해보겠습니다 ㅎㅎ

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    Optional<Category> findByName(String name);
```
사용하지 않는 추상메서드인 것 같아요

KJunseo commented on backend/src/test/java/com/handong/rebon/unit/category/CategoryTest.java:
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class CategoryTest {
```
도메인 단에 새로운 로직이 추가된 것 같아요!(중복검사나, 자식 카테고리 더하는 메서드 등)
도메인 단위 테스트는 추가해주는게 좋을 것 같아요!

KJunseo reviewed (changes requested):
빠르게 반영하셨네요 고생하셨습니다!
조금만 더 같이 보면 좋을 것 같은 부분들이 보여서 리뷰 조금 남겼습니다!

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/member/domain/content/ShopContent.java:
```diff
@@ -1,4 +1,4 @@
-package com.handong.rebon.shop.domain.content;
+package com.handong.rebon.member.domain.content;
 
```
패키지 구조가 잘못된 거 아닌가요??
다른 부분도 확인하시면 좋을거 같아요!!

gomster96 replied:
음 제가 이부분 잘못한거 같습니다..! 제대로 꼬여 버렸네요! 근데 이거 지금 건드려서 원래 있던 위치로 다시 옮기면 이전에 Shop작업했던 분이랑 꼬일 거 같아서.... 제꺼 합쳐지면 그때 따로 issue 파서 삭제하고 다시 합치도록 하겠습니다..>!! 크라켄 이상하게 쓰다가 꼬여버려서 미안합니다..!

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,33 @@
 
-    @OneToMany(mappedBy = "parent")
-    private List<Category> categories;
+    @OneToMany(mappedBy = "parent", cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
+    private List<Category> categories = new ArrayList<>();
```
CascadeType.ALL을 사용하는 건 별로인가요??

PyeongGangKim reviewed (changes requested):
리뷰 남겼습니다!! 고생하셨습니다!

gomster96 committed:
refactor[#4] : 리뷰내용 반영 리팩토링

- Category를 생성할 때 결국 이름으로만 생성하므로 굳이 builder를 쓰지 않아도 될것이라고 생각하여, builder를 제거해주고, 생성자를 새로 만들어 주었다. 
- shopCateogories 컬렉션 필드를 new로 생성하여 두는것이 좋기때문에 반영하였다.
- Categories에서 getter를 응답값의 역할로만 사용하게 하기위해 Category에 method(isSameName)를 만들어 다른 역할의 getter 사용을 줄였다.

gomster96 committed:
refactor[#4] : 사용하지 않는 메서드 제거

gomster96 committed:
test[#4] : 도메인 단위 테스트 추가

아래의 도메인 단위 테스트를 추가했다.
- 부모가 있는 카테고리를 생성
- 다른 부모의 카테고리는 중복된 이름으로 생성 가능
- 같은 부모의 카테고리는 중복된 이름으로 생성 불가

gomster96 committed:
refactor[#4] : 테스트 코드 Displayname 변경

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/application/CategoryService.java:
```diff
@@ -0,0 +1,50 @@
+
+        Category newCategory = new Category(categoryName);
+
+        categoryRepository.save(newCategory);
+        return newCategory.getId();
```
밑에 메서드는 반영되어있는데 이 부분 반영이 안되어있네요 ㅠㅠ

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,33 @@
+    @OneToMany(mappedBy = "parent", cascade = {CascadeType.PERSIST, CascadeType.REMOVE})
+    private List<Category> categories = new ArrayList<>();
+
+    public void addChild(Category category) {
+        this.getCategories().add(category);
```
이 부분이 여전히 getter를 쓰고 있네요

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Categories.java:
```diff
@@ -1,12 +1,33 @@
+        this.getCategories().add(category);
+    }
+
+    public void checkDuplicateCategory(Category category) {
+        this.categories.forEach(child -> checkSameName(child, category));
+    }
+
+    private void checkSameName(Category child, Category category) {
+        String newCategoryName = category.getName();
+        if (child.IsSameName(newCategoryName)) {
+            throw new CategoryExistException();
+        }
```
굳이 2개의 메서드로 분리할 필요가 없어보여요. category의 이름을 get해서 비교하는 것 보다, 메세지를 호출해서 비교하는게 좋아보이네요.

예시) 
```java
    // Categories.java
    public void checkDuplicateCategory(Category category) {
        this.categories.forEach(child -> child.validatesSame(category));
    }

```

```java
    // Category.java
    public void validatesSame(Category category) {
        if (this.name.equals(category.name)) {
            throw new CategoryExistException();
        }
    }
```

gomster96 replied:
이거 좀... 제가 왜 이렇게 짰는지 모르겠네요. 정신이 없었나봐요... 죄송합니다! 잘 가르쳐주셔서 감사해요. 사실 준서씨가 의도한 대로 짜려고 했던건데,,,? 왜 이렇게 만들었는지 .... 흠 아무튼 감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/Category.java:
```diff
@@ -1,23 +1,56 @@
+    public void addChildCategory(Category category) {
+        this.children.checkDuplicateCategory(category);
+        this.children.addChild(category);
+        category.connectParent(this);
```
이 부분을 메서드로 분리하신 이유가 있을까요? 현재 addChildCategory 내부에서만 쓰이고 메서드 자체도 한 줄이라 addChildCategory 안에서 처리하는건 어떨까요? 

원하는 것이 부모 카테고리에 자식 카테고리를 등록하면 자동으로 그 자식 카테고리의 부모를 현재 카테고리로 설정하는 것인데 connectParent()가 public으로 열려있어서 다른 용도로도 사용될 수 있을 것 같네요

KJunseo reviewed (changes requested):
한 번만 더 확인부탁드립니다! 다시 요청해주세요

gomster96 committed:
refactor[#4] : 리뷰내용 반영 리팩토링

gomster96 committed:
refactor[#4] : 도메인 변경

- 카테고리를 최상단 패키지로 위치시켰다.
- 기존에 member에 실수로 shop에 해당하는 도메인을 넣는 등 좀 꼬였는데 머지할 떄 편하게 하도록 바꾸었다.

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/category/domain/category/ShopCategory.java:
```diff
@@ -0,0 +1,19 @@
+package com.handong.rebon.category.domain.category;
+
+import com.handong.rebon.member.domain.item.Shop;
+
```
이 부분에서 error로 인해 코드가 돌아가지 않습니다!

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/review/domain/Review.java:
```diff
@@ -3,12 +3,12 @@
 import java.util.List;
 import javax.persistence.*;
 
-import com.handong.rebon.member.domain.Member;
+import com.handong.rebon.member.Member;
 import com.handong.rebon.review.domain.content.ReviewContent;
 import com.handong.rebon.review.domain.content.ReviewImages;
 import com.handong.rebon.review.domain.content.ReviewScore;
 import com.handong.rebon.review.domain.empathy.Empathy;
-import com.handong.rebon.shop.domain.item.Shop;
+import com.handong.rebon.member.domain.item.Shop;
 
 @Entity
 public class Review {
```
이 부분도 오류가 납니다!

PyeongGangKim reviewed (changes requested):
고생하셨습니다!! 올리시기 전에, 테스트를 한번 돌리고 올리시면 실수를 방지하실 수 있으실 거 같아요!!

gomster96 committed:
fix: resolve conflicts

gomster96 committed:
refactor[#4] : 패키지 이동

- 기존에 패키지 경로가 프로젝트 전체 구성과 맞지 않았기 때문에, 일관성 유지를 위하여 옮겨주었다.

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public void addChildCategory(Category category) {
+        this.children.checkDuplicateCategory(category);
+        this.children.addChild(category);
+        category.parent = this;
```
checkDuplicateCategory(category)를 addChild 내부에서 하는건 어떤가요??
```java
 public void addChildCategory(Category category) {
        this.children.addChild(category);
        category.parent = this;
}
```
```java
    public void addChild(Category category) {
        checkDuplicateCategory(category);
        this.categories.add(category);
    }
```

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Categories.java:
```diff
@@ -0,0 +1,26 @@
+import javax.persistence.OneToMany;
+
+import lombok.Getter;
+
+
```
띄어쓰기 2줄!

KJunseo reviewed (approved):
리뷰를 많이 드린것 같은데 고생많으셨습니다!
이번 pr은 여기서 approve하겠습니다.

++ 패키지 구조 중 /com/handong/rebon/menu는 삭제해도 될거같네요

추가적으로 하나만 의견을 남겼는데 한 번 확인해주세요!

PyeongGangKim reviewed (approved):
고생많으셨습니다!!

gomster96 committed:
refactor[#4] : 리뷰내용 반영
